src/scripttabs.cpp

```diff
@@ -108,7 +108,7 @@ TYPE_SYMBOL asTypeTable[] =
 	{ "",	(INTERP_TYPE)ST_POINTER_STRUCTSTAT, AT_SIMPLE, NULL, NULL },	//for NULLSTRUCTURESTAT
 
 	/* This final entry marks the end of the type list */
-	{ "END OF TYPE LIST", (INTERP_TYPE)NULL,		AT_SIMPLE,	NULL,	NULL },
+	{ "END OF TYPE LIST", (INTERP_TYPE)0,		AT_SIMPLE,	NULL,	NULL },
```

Comment:
As for the code still being broken, with regards to your first point: casting/converting enums is not undefined behaviour it is very well-defined behaviour (given that both the C and C++ specs fully specify how their numerical value gets established).

It's only allowed if the original value is within the range of the enumeration values. I've missed that the code explicitly takes care of that, so you're right. Which doesn't look good though is that 0 casts to VAL_BOOL which is legal INTERP_TYPE value and doesn't look good as sentinel value.

    As for your second comment, it's not the use of C arrays, but rather the reliance on sentinel values (zero) to detect end-of-data that's the problem. Using begin() == array, end() == array + sizeof(array[0]) would work perfectly.

Perhaps you've meant end() == array + sizeof(array)/sizeof(array[0]) - yes, but with that you'll have to always pass and store length or end along with the array pointer.

    In fact that would be better than any C++ container until a container that's fully initializeable by an initializer_list as constexpr comes into existence. The static array has the advantage of everything being able to live in .rodata. (Of course right now it doesn't due to const incorrectness).

I'd argue better as long as it's inconvenient and unsafe to use and doesn't bring any performance improvements (which I don't think is the case here). I'd just convert it to const std::vector<> and pass it by reference to consumers.